packages/flutter/lib/src/widgets/framework.dart

```diff
@@ -3116,6 +3159,11 @@ abstract class Element extends DiagnosticableTree implements BuildContext {
         && depth != null
         && _active
         && Widget.canUpdate(widget, newWidget));
+    assert(() {
```

Comment:
Why do we need to wait with removing the reservation until "update"? Why can we not just call `GlobalKey._debugRemoveFor(this, child);` from "forgetChild" directly?

Reply:
if we only forget child without rebuilding, it is consider globalkey duplication.
The forgetChild is called when globalkey reparent, we want to make sure the slot gets rebuild. This also depends on update method will always clean up child that has been forgotten; otherwise, clearing the _debugForgottenChildrenwithGlobalKey in update method does not make sense. I will update the document.

Reply:
Please add a comment to this effect in the code. People two years from now are going to try to understand what this all does and they won't remember any of this.

And by people I mean me.

Reply:
.. and me :)

Thanks for adding the comment!